libmamba/include/mamba/core/package_download.hpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2023, QuantStack and Mamba Contributors
+//
+// Distributed under the terms of the BSD 3-Clause License.
+//
+// The full license is in the file LICENSE, distributed with this software.
+
+
+#ifndef MAMBA_CORE_PACKAGE_DOWNLOAD_HPP
+#define MAMBA_CORE_PACKAGE_DOWNLOAD_HPP
+
+#include <future>
+#include <memory>
+#include <set>
+#include <string>
+#include <tuple>
+#include <vector>
+
+#include "fetch.hpp"
+#include "mamba_fs.hpp"
+#include "package_cache.hpp"
+#include "progress_bar.hpp"
+#include "thread_utils.hpp"
+
+namespace mamba
+{
+    class ChannelContext;
+
+    class PackageDownloadExtractTarget
+    {
+    public:
+
+        enum VALIDATION_RESULT
+        {
+            UNDEFINED = 0,
+            VALID = 1,
+            SHA256_ERROR,
+            MD5SUM_ERROR,
+            SIZE_ERROR,
+            EXTRACT_ERROR
+        };
+
+        PackageDownloadExtractTarget(const PackageInfo& pkg_info, ChannelContext& channel_context);
+
+        void write_repodata_record(const fs::u8path& base_path);
+        void add_url();
+        bool finalize_callback(const DownloadTarget& target);
+        bool finished();
+        void validate();
+        bool extract();
+        bool extract_from_cache();
+        bool validate_extract();
+        const std::string& name() const;
+        std::size_t expected_size() const;
+        VALIDATION_RESULT validation_result() const;
+        void clear_cache() const;
+
+        DownloadTarget* target(MultiPackageCache& cache);
+
+        std::exception m_decompress_exception;
```

Comment:
BTW do you know what this object is and why it's public?

Reply:
I have not yet worked with that code. 
M guess is catching the exception is thread to rethrow in the main one.